core/src/main/scala/org/bitcoins/core/psbt/PSBT.scala

```diff
@@ -0,0 +1,1870 @@
+    type RecordType = InputPSBTRecord.RedeemScript
+  }
+
+  final case object WitnessScriptKeyId extends PSBTInputKeyId {
+    override val byte: Byte = 0x05.byteValue
+    type RecordType = InputPSBTRecord.WitnessScript
+  }
+
+  final case object BIP32DerivationPathKeyId extends PSBTInputKeyId {
+    override val byte: Byte = 0x06.byteValue
+    type RecordType = InputPSBTRecord.BIP32DerivationPath
+  }
+
+  final case object FinalizedScriptSigKeyId extends PSBTInputKeyId {
+    override val byte: Byte = 0x07.byteValue
+    type RecordType = InputPSBTRecord.FinalizedScriptSig
+  }
+
+  final case object FinalizedScriptWitnessKeyId extends PSBTInputKeyId {
+    override val byte: Byte = 0x08.byteValue
+    type RecordType = InputPSBTRecord.FinalizedScriptWitness
+  }
+
+  final case object ProofOfReservesCommitmentKeyId extends PSBTInputKeyId {
+    override val byte: Byte = 0x09.byteValue
+    type RecordType = InputPSBTRecord.ProofOfReservesCommitment
+  }
+
+  final case object UnknownKeyId extends PSBTInputKeyId {
+    override val byte: Byte = Byte.MaxValue
+    type RecordType = InputPSBTRecord.Unknown
+  }
+}
+
+sealed trait PSBTOutputKeyId extends PSBTKeyId {
+  type RecordType <: OutputPSBTRecord
+}
+
+object PSBTOutputKeyId extends PSBTKeyIdFactory[PSBTOutputKeyId] {
+
+  override def unknownKey: PSBTOutputKeyId = UnknownKeyId
+
+  override def fromByte(byte: Byte): PSBTOutputKeyId = byte match {
+    case RedeemScriptKeyId.byte        => RedeemScriptKeyId
+    case WitnessScriptKeyId.byte       => WitnessScriptKeyId
+    case BIP32DerivationPathKeyId.byte => BIP32DerivationPathKeyId
+    case _: Byte                       => UnknownKeyId
+  }
+
+  final case object RedeemScriptKeyId extends PSBTOutputKeyId {
+    override val byte: Byte = 0x00.byteValue
+    type RecordType = OutputPSBTRecord.RedeemScript
+  }
+
+  final case object WitnessScriptKeyId extends PSBTOutputKeyId {
+    override val byte: Byte = 0x01.byteValue
+    type RecordType = OutputPSBTRecord.WitnessScript
+  }
+
+  final case object BIP32DerivationPathKeyId extends PSBTOutputKeyId {
+    override val byte: Byte = 0x02.byteValue
+    type RecordType = OutputPSBTRecord.BIP32DerivationPath
+  }
+
+  final case object UnknownKeyId extends PSBTOutputKeyId {
+    override val byte: Byte = Byte.MaxValue
+    type RecordType = OutputPSBTRecord.Unknown
+  }
+}
+
+sealed trait PSBTMap[+RecordType <: PSBTRecord] extends NetworkElement {
+  require(elements.map(_.key).groupBy(identity).values.forall(_.length == 1),
+          "All keys must be unique.")
+
+  def elements: Vector[RecordType]
+
+  def bytes: ByteVector =
+    elements
+      .sortBy(_.key)
+      .foldLeft(ByteVector.empty)(_ ++ _.bytes) :+ PSBTMap.separatorByte
+
+  protected def getRecords[KeyIdType <: PSBTKeyId](
+      key: KeyIdType,
+      keyIdFactory: PSBTKeyIdFactory[KeyIdType]): Vector[key.RecordType] = {
+    elements
+      .filter(element => keyIdFactory.fromByte(element.key.head) == key)
```

Comment:
Could we just use `.collect { } ` here and avoid the cast?

Reply:
@nkohen

Reply:
Sadly no, using collect with `key.RecordType` causes a warning message because in fact either way this is done, there is type erasure and nothing is actually happening at run time.